### DISCUSSION TITLE
Connect-PnPOnline suddenly at times getting "Could not load file or assembly 'Microsoft.Identity.Client, Version=4.50.0.0'" on script that has been working

### DISCUSSION THREAD

**Discussion by jdenicola-pa** [pnp/powershell#3216] [Q&A]
I have a script that does the following command:

`Connect-PnPOnline -Url $site -Tenant $tenant -ClientId $appClientID -CertificatePath $cert -CertificatePassword $(ConvertTo-Securestring -string $certPassword -AsPlainText -Force)
`

I run this script from the Task Scheduler. It is failing with:

```
Connect-PnPOnline: Could not load file or assembly 'Microsoft.Identity.Client, Version=4.50.0.0, Culture=neutral, PublicKeyToken=0a613f4dd989e8ae'. Could not
find or load a specific file. (0x80131621)
```

I read somewhere else that this could be affected by the Azure Powershell object (which I did not have installed), so I installed them.

Now here is the odd thing:

- If I run the above command from the Terminal pane of Visual Studio Code, I get the error
- If I run the above command from my script in the Task Scheduler, I get the error
- If I open a Powershell command window (c:\program files\PowerShell\7\pwsh.exe), I do **not** get the error

In VS Code, $PSVersionTable says PSVersion = 7.3.4
From the Task Scheduler, PSVersion = 7.3.4
From the Powershell command window, PSVersion = 7.3.4

Get-InstalledModule -Name PnP.Powershell returns version 2.1.1

All are executing as me.

I then tried;

`Install-Module -Name Microsoft.Identity.Client -RequiredVersion 4.50.0.0`

Now on Connect-PnPOnline, I get:

`Could not load type 'Microsoft.Extensions.DependencyInjection.ServiceCollection' from assembly 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=5.0.0.0`

Still works from the Powershell command window I open, but does not from VS Code or the Task Scheduler.

Why would it fail sometimes and work other times? What might be the cause?

**Comment by Korez73** [comment]
I am also getting that second error  
`Could not load type 'Microsoft.Extensions.DependencyInjection.ServiceCollection' from assembly 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=5.0.0.0` in VSCode.  Running from powershell also works fine for me.

**Reply by jdenicola-pa** [reply]
Here is something very strange that I found.

I had the Connect-PnPOnline inside of a Powershell function in my script.

I moved it to outside of the function, so it is in the "main" code that executes when the script starts, and it works there! I move it back inside the function and it fails.

For example:

```
function MyFunction
{
     # Fails
     Connect-PnPOnline -various - options

     # Do some things with SharePoint
     return $value
}

$val = MyFunction
```


where as:

```
function MyFunction
{

     # Do some things with SharePoint
     return $value
}

# Succeeds!
Connect-PnPOnline -various - options

$val = MyFunction
```

I ended up removing Microsoft.Identity.Client altogether using:

Uninstall-Module -name Microsoft.Identity.Client -AllVersions

I can't explain it.

**Comment by AndersRask** [comment]
The issue with colliding assemblies should now be fixed!
https://github.com/pnp/powershell/issues/3207

**Comment by Soepgroenten** [comment]
I suddenly also experienced this error:
`"Connect-PnPOnline: Could not load type 'Microsoft.Extensions.DependencyInjection.ServiceCollection' from assembly 'Microsoft.Extensions.DependencyInjection.Abstractions"`

So I ended up with this thread which in someway pointed me in finding a workaround, so I decided to share it with you as this is beyond all logic.
I got a script that doesn't use any other modules so the only one was the "PnP.PowerShell"module.
I've resolved this by simply putting the "Import-Module PnP.PowerShell" and the "Connect-PnPOnline" on top of all code which works. 

Yes it results in my script not being chronologically anymore but at least I got my scripts working again.

As a Bonus I previously also struggled with the colliding assemblies issues, but I got that one solved by writing this (also copied my comments around it :-) ):
```
    # Perform some prereq checks (Like having the MSOL Sharepoint module installed)
        # Check if PnP Powershell is installed, if not --> Install it to the running user's ps modules library to avoid the need to run this script as admin (Documentation: https://learn.microsoft.com/en-us/powershell/sharepoint/sharepoint-pnp/sharepoint-pnp-cmdlets?view=sharepoint-ps)
        If ((Get-InstalledModule PnP.PowerShell -ErrorAction SilentlyContinue) -eq $null){
            Install-Module PnP.PowerShell -Scope CurrentUser -Force -Confirm:$false
            }
        # Delete the Microsoft.Graph.Core.dll file out of the install dir of pnp.powershell module --> This is a workaround to make the script work the file will be recreated at every update so I've put this delete action in the script to make sure the script doesnt break.
            # Gather the installdir for the pnp-powershell module
                $PnPModuleDirectory = (Get-Module PnP.PowerShell -ListAvailable)[0].ModuleBase
            # If the Microsoft.Graph.Core.dll file is present --> Delete it!
            if (Test-Path -Path "$PnPModuleDirectory\Core\Microsoft.Graph.Core.dll"){
                # The actual remove action
                    Remove-Item -Path ($PnPModuleDirectory + "\Core\Microsoft.Graph.Core.dll") -Force -Verbose
            }
```

This installs the module if missing, and remove the .dll file which is was causing the issue. Actually I expected issues with some parts of the module, but I didn't found any so far.

Please take note that these simple fixes ar 100% workarounds, my suggestion is to regularly check if the issues got fixed and make your script work as it should be.

Hopefully these simple fixes helps someone continuing their business without a hassling further :-)
